nav2_system_tests/src/system/nav_to_pose_tester_node.py

```diff
@@ -404,7 +405,7 @@ def main(argv=sys.argv[1:]):
 
     for tester in testers:
         passed = run_all_tests(tester)
-        if not passed:
+        if str(passed) != str(args.expect_failure):
```

Comment:
Don't compare strings, these should be tested as bools (as well below)

Reply:
I used strings because `passed != args.expect_failure` didn't work as expected, even if `args.expect_failure` is evaluated correctly. 

What about `passed != eval(args.expect_failure)` ?  It definitely works (I've already test it), but I don't know if it's the best way.

Reply:
You might want to error check before `eval` but that's generally an OK strategy here since this is a test. I'd just verify the string is exactly `True` or `False` and if not, log an error because eval is going to probably fail

Reply:
I put the check on `eval()` argument in `get_testers(args)`, so that I can use the `error_msg` already define.

    if args.robot:
        # Requested tester for one robot
        init_x, init_y, final_x, final_y = args.robot[0]
        tester = NavTester(
            initial_pose=fwd_pose(float(init_x), float(init_y)),
            goal_pose=fwd_pose(float(final_x), float(final_y)))
        # Check if --expect_failure is True or False
        if args.expect_failure != 'True' and args.expect_failure != 'False':
            tester.error_msg("-e flag must be set True or False only.")
            exit(1)
        tester.info_msg(
            'Starting tester, robot going from ' + init_x + ', ' + init_y +
            ' to ' + final_x + ', ' + final_y + '.')
        testers.append(tester)
        return testers

    # Requested tester for multiple robots
    for robot in args.robots:
        namespace, init_x, init_y, final_x, final_y = robot
        tester = NavTester(
            namespace=namespace,
            initial_pose=fwd_pose(float(init_x), float(init_y)),
            goal_pose=fwd_pose(float(final_x), float(final_y)))
        # Check if --expect_failure is True or False
        if args.expect_failure != 'True' and args.expect_failure != 'False':
            tester.error_msg("-e flag must be set True or False only.")
            exit(1)
        tester.info_msg(
            'Starting tester for ' + namespace +
            ' going from ' + init_x + ', ' + init_y +
            ' to ' + final_x + ', ' + final_y)
        testers.append(tester)
    return testers

Reply:
I create a new **NavTester** method:
```
def check_args(self, expect_failure: str):
  # Check if --expect_failure is True or False
  if expect_failure != 'True' and expect_failure != 'False':
    self.error_msg("-e flag must be set True or False only.")
    exit(1)
```

then call it in `get_testers(args)`:

```
def get_testers(args):
    testers = []

    if args.robot:
        # Requested tester for one robot
        init_x, init_y, final_x, final_y = args.robot[0]
        tester = NavTester(
            initial_pose=fwd_pose(float(init_x), float(init_y)),
            goal_pose=fwd_pose(float(final_x), float(final_y)))
        tester.check_args(args.expect_failure)
        tester.info_msg(
            'Starting tester, robot going from ' + init_x + ', ' + init_y +
            ' to ' + final_x + ', ' + final_y + '.')
        testers.append(tester)
        return testers

    # Requested tester for multiple robots
    for robot in args.robots:
        namespace, init_x, init_y, final_x, final_y = robot
        tester = NavTester(
            namespace=namespace,
            initial_pose=fwd_pose(float(init_x), float(init_y)),
            goal_pose=fwd_pose(float(final_x), float(final_y)))
        tester.check_args(args.expect_failure)
        tester.info_msg(
            'Starting tester for ' + namespace +
            ' going from ' + init_x + ', ' + init_y +
            ' to ' + final_x + ', ' + final_y)
        testers.append(tester)
    return testers
```

Reply:
Or you could just make check args an independent function from the tester and have it return the bool for T/F while its checking (to use then later rather than eval in the `main()`)

Reply:
What do you think about last commit?

Even if locally everything works fine, CI build shows this result for *test_bt_navigator_with_wrong_init_pose*:

`The test did not generate a result file.`

Reply:
That means the test crashed, failed, or hung, you should run locally and debug.

Reply:
I just checked [CI builds](https://app.circleci.com/pipelines/github/ros-planning/navigation2?branch=pull%2F2420). 

2h ago another build has been triggered somehow, and the error relative to *test_bt_navigator_with_wrong_init_pose* is gone.

![Screenshot from 2021-07-12 21-05-43](https://user-images.githubusercontent.com/24660508/125342717-82bbc080-e355-11eb-9c10-fbd6f86fa203.png)

![Screenshot from 2021-07-12 21-05-31](https://user-images.githubusercontent.com/24660508/125342737-86e7de00-e355-11eb-9bae-69007a246d24.png)

Reply:
I retriggered it just in case it was a transient failure, looks like it was!